### PR TITLE
Updates gcc package to enable compiler offload to AMD GCN.

### PR DESCRIPTION
**Pull request by frobnitzem**
Works for gcc 10 following instructions at [gnu.org](https://gcc.gnu.org/wiki/Offloading).
Tested on Instinct MI60 (Vega 20) system.

### CODE REVIEW

File: var/spack/repos/builtin/packages/gcc/package.py [hunk 1]
```diff
@@ -124,6 +127,9 @@ class Gcc(AutotoolsPackage, GNUMirrorPackage):
     depends_on('binutils~libiberty', when='+binutils', type=('build', 'link', 'run'))
     depends_on('zip', type='build', when='languages=java')
     depends_on('cuda@:10', when='+nvptx')
+    #depends_on('llvm-amdgpu', when='+amdgcn') # not binary compatible with llvm9 fmt.
+    # FIXME: there's no way to specify this in spack, it's a variant, not a dep
+    #depends_on('languages=lto', when='+amdgcn')
```
Comment (frobnitzem) [inline var/spack/repos/builtin/packages/gcc/package.py]: Note: The addition of link time optimization is a suggestion from https://gcc.gnu.org/wiki/Offloading, not a requirement.

File: var/spack/repos/builtin/packages/gcc/package.py [hunk 2]
```diff
@@ -124,6 +127,10 @@ class Gcc(AutotoolsPackage, GNUMirrorPackage):
     depends_on('binutils~libiberty', when='+binutils', type=('build', 'link', 'run'))
     depends_on('zip', type='build', when='languages=java')
     depends_on('cuda@:10', when='+nvptx')
+    # depends_on('llvm-amdgpu', when='+amdgcn')
+    # not binary compatible with llvm9 fmt.
+    # FIXME: there's no way to specify this in spack, it's a variant, not a dep
+    # depends_on('languages=lto', when='+amdgcn')
```
Comment (michaelkuhn) [inline var/spack/repos/builtin/packages/gcc/package.py:133]: Maybe add a conflict somehow to make sure that lto is enabled?

File: var/spack/repos/builtin/packages/gcc/package.py [hunk 3]
```diff
@@ -134,6 +141,18 @@ class Gcc(AutotoolsPackage, GNUMirrorPackage):
              destination='newlibsource',
              when='+nvptx',
              fetch_options=timeout)
+    resource(name='newlib-amdgcn',
+             url='ftp://sourceware.org/pub/newlib/newlib-3.3.0.tar.gz',
+             sha256='58dd9e3eaedf519360d92d84205c3deef0b3fc286685d1c562e245914ef72c66',
+             destination='newlibsource',
+             when='+amdgcn',  # TODO: Is there an or-selection syntax?
+             fetch_options=timeout)  # to combine with newlib, above
```
Comment (michaelkuhn) [inline var/spack/repos/builtin/packages/gcc/package.py:149]: What will happen if both nvptx and amdgcn are enabled? They seem to extract into the same destination.

File: var/spack/repos/builtin/packages/gcc/package.py [hunk 4]
```diff
@@ -134,6 +141,18 @@ class Gcc(AutotoolsPackage, GNUMirrorPackage):
...
+    resource(name='llvm-amdgcn',
+             url='https://github.com/llvm/llvm-project/archive/llvmorg-9.0.0.tar.gz',
+             sha256='7807fac25330e24e9955ca46cd855dd34bbc9cc4fdba8322366206654d1036f2',
+             destination='llvm-source',
+             when='+amdgcn',
+             fetch_options=timeout)
```
Comment (michaelkuhn) [inline var/spack/repos/builtin/packages/gcc/package.py:155]: Isn't this just llvm? Why can't we have a dependency on it?

File: var/spack/repos/builtin/packages/gcc/package.py [hunk 5]
```diff
@@ -209,6 +229,8 @@ class Gcc(AutotoolsPackage, GNUMirrorPackage):
     conflicts('languages=obj-c++', when='+nvptx')
     # NVPTX build disables bootstrap
     conflicts('+binutils', when='+nvptx')
+    # FIXME: the languages above probably also conflict with amdgcn
+    conflicts('+binutils', when='+amdgcn')
```
Comment (michaelkuhn) [inline var/spack/repos/builtin/packages/gcc/package.py:233]: I guess both of these should now conflict with +bootstrap.

**Review by michaelkuhn — COMMENTED** [review]
Sorry, I somehow completely missed this.

I haven't tried building it yet, a few comments/questions, though.

File: var/spack/repos/builtin/packages/gcc/package.py (hunk 4 above)
Comment (frobnitzem) [reply var/spack/repos/builtin/packages/gcc/package.py:155]: According to gcc 10's build instructions, it's only the binutils part of llvm, with a special mod I think.

**Comment by frobnitzem** [comment]
GCC 11 is now released.

**Pull request closed without merging**